### DISCUSSION TITLE
Question about asRegExp

### DISCUSSION THREAD

**Discussion by inottn** [webpack/webpack#18873] [General]
The [`asRegExp`](https://github.com/webpack/webpack/blob/93743d233ab4fa36738065ebf8df5f175323b906/lib/ModuleFilenameHelpers.js#L103-L110) function converts the parameter `test` (if it is a string) into a regular expression, which is then used by the [`matchPart`](https://github.com/webpack/webpack/blob/93743d233ab4fa36738065ebf8df5f175323b906/lib/ModuleFilenameHelpers.js#L338-L345) function for matching. Compared to directly using a string for matching, what problem does it solve? I would greatly appreciate any answers.

**Comment by alexander-akait** [comment]
Using `test: ['test', /test/]`

**Reply by inottn** [reply]
Thank you for your response. I noticed that 'test' is being converted to /^test/. Could we avoid this conversion and use str.startsWith to match the result directly instead?

**Reply by alexander-akait** [reply]
I think so, but `startsWith` worked a little slower than `RegExp` before, need to check

**Reply by inottn** [reply]
Sorry for the late reply; I've been on vacation recently 😄. Here are the online [benchmark results](https://www.measurethat.net/Benchmarks/Show/14284/0/js-regex-vs-startswith-vs-indexof-simpler#latest_results_block), which conclude that the fastest method is indexOf, while RegExp is the slowest. Since it doesn't include the step of converting a string to a regex, RegExp is actually slower. I tested it on different Node versions and got the same results. Maybe we can use indexOf?

**Reply by alexander-akait** [reply]
@inottn Great job, yeah we need to use `indexOf`, even more I think we should set up a eslint rule that which prohibits to use `startsWith` and check `regex start` our code base (but we can do this in a few prs), do you want to send a pr?

**Reply by inottn** [reply]
I can give it a try. 😄
